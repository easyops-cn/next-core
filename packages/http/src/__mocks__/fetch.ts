import "whatwg-fetch";

let returnValue: any;

export const fetch = jest.fn(
  () => returnValue || Promise.resolve(new Response("{}"))
);
export const __setReturnValue = (value: any): void => {
  returnValue = value;
};
