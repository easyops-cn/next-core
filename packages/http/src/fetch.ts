export const fetch = (
  input: RequestInfo,
  init?: RequestInit
): Promise<Response> => {
  // Support older browsers which default credentials is "omit".
  // Ref https://github.com/whatwg/fetch/pull/585
  const req = new Request(
    window.__INJECTED_PUBLIC_PATH_BY_QIANKUN__
      ? typeof input === "string"
        ? new URL(input, window.__INJECTED_PUBLIC_PATH_BY_QIANKUN__).toString()
        : {
            ...input,
            url: new URL(input.url, window.__INJECTED_PUBLIC_PATH_BY_QIANKUN__).toString(),
          }
      : input,
    Object.assign<RequestInit, RequestInit | undefined>(
      {
        credentials: window.__INJECTED_PUBLIC_PATH_BY_QIANKUN__ ? "include" : "same-origin",
      },
      init
    )
  );

  return self.fetch(req);
};
