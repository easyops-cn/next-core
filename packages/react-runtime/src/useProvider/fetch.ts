import { FetchArgs } from "./useProviderTypes.js";
import { fetchByProvider } from "@next-core/runtime";

const cacheMap: Map<string, Promise<unknown>> = new Map();

function isObj(v: any): boolean {
  return typeof v === "object" && v !== null && !Array.isArray(v);
}

function isString(v: any): v is string {
  return typeof v === "string";
}
function buildSortedCacheKey(provider: string, ...args: any): string {
  const sortObj = (obj: Record<string, any>) =>
    Object.keys(obj)
      .sort()
      .map((k) => ({ [k]: (obj as any)[k] }));
  try {
    const sortedArgs = isObj(args)
      ? sortObj(args)
      : (args as Record<string, any>[]).map((arg) =>
          isString(arg) ? arg : sortObj(arg)
        );

    return JSON.stringify({
      provider,
      args: sortedArgs,
    });
  } catch (e) {
    return JSON.stringify({
      provider,
      args,
    });
  }
}

export default async function fetch<TData>(
  provider: string,
  cache: boolean,
  args: FetchArgs
): Promise<TData> {
  let promise: Promise<TData>;
  // eslint-disable-next-line @typescript-eslint/ban-ts-comment
  // @ts-ignore
  const cacheKey = buildSortedCacheKey(provider, ...args);
  !cache && cacheMap.has(cacheKey) && cacheMap.delete(cacheKey);

  if (cacheMap.has(cacheKey)) {
    promise = cacheMap.get(cacheKey) as Promise<TData>;
  } else {
    promise = (() => {
      return fetchByProvider(provider, args, {cache: cache ? "default" : "reload"}) as Promise<TData>;
    })();

    cache && cacheMap.set(cacheKey, promise);
  }

  return promise;
}
