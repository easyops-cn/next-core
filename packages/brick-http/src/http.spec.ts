// @ts-ignore mocked
import { fetch, __setReturnValue } from "./fetch";
import {
  http,
  HttpOptions,
  HttpFetchError,
  HttpParseError,
  HttpResponseError,
  HttpAbortError,
} from "./";

jest.mock("./fetch");

const spyOnFetch = fetch as jest.Mock;

type TestItem =
  | ["GET" | "DELETE" | "HEAD", string, [HttpOptions?]]
  | ["POST" | "PUT" | "PATCH", string, [any?, HttpOptions?]];

// Hack for snapshot testing.
(FormData.prototype as any).toJSON = function () {
  return "<FormData> " + JSON.stringify(Array.from(this.entries()));
};
(URLSearchParams.prototype as any).toJSON = function () {
  return "<URLSearchParams> " + this.toString();
};

describe("http", () => {
  http.interceptors.request.use(
    (config) => {
      return config;
    },
    (error) => {
      return Promise.reject(error);
    }
  );

  http.interceptors.response.use(
    (response) => {
      return response.config.options?.observe === "response"
        ? response
        : response.data;
    },
    (error) => {
      return Promise.reject(error.error);
    }
  );

  afterEach(() => {
    http.enableCache(false);
    spyOnFetch.mockClear();
  });

  const formData = new FormData();
  formData.append("for", "the throne");

  const batchTests: TestItem[] = [
    ["GET", "http://example.com", []],
    ["POST", "http://example.com", []],
    ["PUT", "http://example.com", []],
    ["PATCH", "http://example.com", []],
    ["DELETE", "http://example.com", []],
    ["HEAD", "http://example.com", []],
    [
      "GET",
      "http://example.com",
      [{ params: { lang: "zh-CN", type: null, category: undefined } }],
    ],
    [
      "GET",
      "http://example.com",
      [{ params: new URLSearchParams("lang=zh-CN") }],
    ],
    [
      "GET",
      "http://example.com",
      [
        {
          params: { lang: ["zh-CN", "en-US"], type: null, category: undefined },
        },
      ],
    ],
    [
      "GET",
      "http://example.com/for-good?token=secret",
      [{ params: { lang: "zh-CN", type: null, category: undefined } }],
    ],
    ["POST", "http://example.com", [{ for: "good" }]],
    [
      "POST",
      "http://example.com/for-good",
      [
        { for: "the throne" },
        { params: { lang: "zh-CN", type: null, category: undefined } },
      ],
    ],
    [
      "POST",
      "http://example.com/for-good",
      [
        { for: "the throne" },
        {
          params: new URLSearchParams("lang=zh-CN"),
        },
      ],
    ],
    [
      "POST",
      "http://example.com/for-good",
      [
        { for: "the throne" },
        { headers: { "x-requested-with": "XMLHttpRequest" } },
      ],
    ],
    [
      "POST",
      "http://example.com/for-good",
      [
        { for: "the throne" },
        { headers: new Headers({ "x-requested-with": "XMLHttpRequest" }) },
      ],
    ],
    ["POST", "http://example.com/for-good", ["for=the-throne"]],
    ["POST", "http://example.com/for-good", [formData]],
  ];
  it.each(batchTests)(
    "%s %s with %j should work",
    async (method: string, url: string, args: any[]) => {
      await http[method.toLowerCase() as "get"](url, ...args);
      expect(spyOnFetch.mock.calls[0]).toMatchSnapshot();
    }
  );

  it.each(batchTests)(
    `requestWithBody or simpleRequest %s %s with %j should work`,
    async (method: string, url: string, args: any[]) => {
      let requestType = "simpleRequest";
      ["POST", "PUT", "PATCH"].includes(method) &&
        (requestType = "requestWithBody");
      await http[requestType as "post"](method, url, ...args);
      expect(spyOnFetch.mock.calls[0]).toMatchSnapshot();
    }
  );

  it("should work with http.request", async () => {
    await http.request("http://example.com/for-good", {
      method: "GET",
    });

    expect(spyOnFetch.mock.calls[0]).toMatchSnapshot();
  });

  it("preview was true and http request should use cache", async () => {
    const mockListen = jest.fn();
    http.on("match-api-cache", mockListen);
    http.enableCache(true);
    http.setClearCacheIgnoreList([
      {
        method: "GET",
      },
      {
        method: "POST",
        uri: "http://example.com",
      },
    ]);
    // get
    await http.get("http://example.com");
    expect(spyOnFetch).toBeCalledTimes(1);

    await http.get("http://example.com");
    expect(spyOnFetch).toBeCalledTimes(1);

    await http.get("http://example.com?v=1");
    expect(spyOnFetch).toBeCalledTimes(2);

    await http.get("http://example.com");
    expect(spyOnFetch).toBeCalledTimes(2);

    // post
    await http.post("http://example.com");
    expect(spyOnFetch).toBeCalledTimes(3);

    await http.post("http://example.com", {
      v: 1,
    });
    expect(spyOnFetch).toBeCalledTimes(4);

    await http.post("http://example.com", {
      v: 1,
    });
    expect(spyOnFetch).toBeCalledTimes(4);

    await http.post("http://example.com", {
      v: 2,
    });
    expect(spyOnFetch).toBeCalledTimes(5);

    await http.get("http://example.com", {
      params: {
        v: 3,
      },
    });
    expect(spyOnFetch).toBeCalledTimes(6);

    await http.get("http://example.com", {
      params: {
        v: 3,
      },
    });
    expect(spyOnFetch).toBeCalledTimes(6);

    await http.get("http://example.com", {
      params: {
        v: 4,
      },
    });
    expect(spyOnFetch).toBeCalledTimes(7);

    expect(mockListen).toHaveBeenLastCalledWith(6);
  });

  it("preview was true and http request should use clearCacheIgnoreList", async () => {
    const mockListen = jest.fn();
    http.on("match-api-cache", mockListen);
    http.enableCache(true);
    http.setClearCacheIgnoreList([
      {
        method: "GET",
      },
      {
        method: "POST",
        uri: ".*cmdb.instance.PostSearch.*",
      },
      {
        method: "POST",
        uri: "api/gateway/data_exchange.olap.Query/api/v1/data_exchange/olap",
      },
    ]);
    // get
    await http.get(
      "api/gateway/cmdb.instance.GetDetail/object/HOST/instance/58fe908324229"
    );
    expect(spyOnFetch).toBeCalledTimes(1);
    await http.post(
      "api/gateway/cmdb.instance.PostSearch/object/HOST/instance/_search",
      {
        fields: { ip: 1 },
      }
    );
    expect(spyOnFetch).toBeCalledTimes(2);
    await http.post(
      "api/gateway/data_exchange.olap.Query/api/v1/data_exchange/olap",
      {
        test: 1,
      }
    );
    expect(spyOnFetch).toBeCalledTimes(3);

    await http.get(
      "api/gateway/cmdb.instance.GetDetail/object/HOST/instance/58fe908324229"
    );
    expect(spyOnFetch).toBeCalledTimes(3);
    await http.post(
      "api/gateway/cmdb.instance.PostSearch/object/HOST/instance/_search",
      {
        fields: { ip: 1 },
      }
    );
    expect(spyOnFetch).toBeCalledTimes(3);
    // 在 clearCacheIgnoreList 名单外，清除缓存
    await http.post(
      "api/gateway/data_exchange.olap.Query/api/v2/data_exchange/olap",
      {
        test: 1,
      }
    );
    expect(spyOnFetch).toBeCalledTimes(4);
    expect(mockListen).toHaveBeenLastCalledWith(10);
  });

  it("should work with getUrlWithParams", () => {
    const result = http.getUrlWithParams("http://example.com/for-good", {
      name: "monkey",
    });
    expect(result).toEqual("http://example.com/for-good?name=monkey");
  });

  it("should work with getBodyAndHeaders", () => {
    const result = http.getBodyAndHeaders("row-data", {});
    const headers = new Headers({
      "content-type": "application/x-www-form-urlencoded",
    });
    expect(result).toEqual({ body: "row-data", headers: headers });
  });

  it("should work with getBodyAndHeaders when provided content-type header", () => {
    const result = http.getBodyAndHeaders('"row-data"', {
      "content-type": "application/json",
    });
    const headers = new Headers({
      "content-type": "application/json",
    });
    expect(result).toEqual({ body: '"row-data"', headers: headers });
  });

  it("should return raw text", async () => {
    __setReturnValue(Promise.resolve(new Response("raw-text")));
    const result = await http.get("http://example.com", {
      responseType: "text",
    });
    expect(result).toBe("raw-text");
  });

  it("should return blob", async () => {
    __setReturnValue(Promise.resolve(new Response(new Blob())));
    const result = await http.get("http://example.com", {
      responseType: "blob",
    });
    expect(result).toBeInstanceOf(Blob);
  });

  it("should return array buffer", async () => {
    __setReturnValue(Promise.resolve(new Response(new ArrayBuffer(1))));
    const result = await http.get("http://example.com", {
      responseType: "arrayBuffer",
    });
    expect(result).toBeInstanceOf(ArrayBuffer);
  });

  it("should throw a HttpFetchError", async () => {
    __setReturnValue(Promise.reject(new Error("mock error")));
    expect.assertions(1);
    try {
      await http.get("http://example.com");
    } catch (e) {
      expect(e).toBeInstanceOf(HttpFetchError);
    }
  });

  it("should throw a HttpResponseError", async () => {
    __setReturnValue(
      Promise.resolve(
        new Response('{"error":"oops"}', {
          status: 500,
        })
      )
    );
    expect.assertions(2);
    try {
      await http.post("http://example.com");
    } catch (e) {
      expect(e).toBeInstanceOf(HttpResponseError);
      expect(e.responseJson).toEqual({
        error: "oops",
      });
    }
  });

  it("should throw a HttpParseError", async () => {
    __setReturnValue(Promise.resolve(new Response("non-json")));
    expect.assertions(1);
    try {
      await http.get("http://example.com");
    } catch (e) {
      expect(e).toBeInstanceOf(HttpParseError);
    }
  });

  it("should throw a HttpAbortError", async () => {
    __setReturnValue(
      Promise.reject(
        new DOMException("The user aborted a request.", "AbortError")
      )
    );
    expect.assertions(1);
    try {
      await http.get("http://example.com");
    } catch (e) {
      expect(e).toBeInstanceOf(HttpAbortError);
    }
  });

  it("should return http response object ", async () => {
    __setReturnValue(
      Promise.resolve(new Response(JSON.stringify({ foo: "bar" })))
    );
    const result = await http.get("http://example.com", {
      observe: "response",
    });
    expect(result).toMatchSnapshot();
  });
});
