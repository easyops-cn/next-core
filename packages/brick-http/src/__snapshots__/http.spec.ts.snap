// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`http DELETE http://example.com with [] should work 1`] = `
Array [
  "http://example.com",
  Object {
    "method": "DELETE",
  },
]
`;

exports[`http GET http://example.com with [] should work 1`] = `
Array [
  "http://example.com",
  Object {
    "method": "GET",
  },
]
`;

exports[`http GET http://example.com with [{"params":"<URLSearchParams> lang=zh-CN"}] should work 1`] = `
Array [
  "http://example.com/?lang=zh-CN",
  Object {
    "method": "GET",
  },
]
`;

exports[`http GET http://example.com with [{"params":{"lang":"zh-CN","type":null}}] should work 1`] = `
Array [
  "http://example.com/?lang=zh-CN&type=&category=",
  Object {
    "method": "GET",
  },
]
`;

exports[`http GET http://example.com with [{"params":{"lang":["zh-CN","en-US"],"type":null}}] should work 1`] = `
Array [
  "http://example.com/?lang=zh-CN&lang=en-US&type=&category=",
  Object {
    "method": "GET",
  },
]
`;

exports[`http GET http://example.com/for-good?token=secret with [{"params":{"lang":"zh-CN","type":null}}] should work 1`] = `
Array [
  "http://example.com/for-good?token=secret&lang=zh-CN&type=&category=",
  Object {
    "method": "GET",
  },
]
`;

exports[`http HEAD http://example.com with [] should work 1`] = `
Array [
  "http://example.com",
  Object {
    "method": "HEAD",
  },
]
`;

exports[`http PATCH http://example.com with [] should work 1`] = `
Array [
  "http://example.com",
  Object {
    "headers": undefined,
    "method": "PATCH",
  },
]
`;

exports[`http POST http://example.com with [] should work 1`] = `
Array [
  "http://example.com",
  Object {
    "headers": undefined,
    "method": "POST",
  },
]
`;

exports[`http POST http://example.com with [{"for":"good"}] should work 1`] = `
Array [
  "http://example.com",
  Object {
    "body": "{\\"for\\":\\"good\\"}",
    "headers": Headers {
      "map": Object {
        "content-type": "application/json",
      },
    },
    "method": "POST",
  },
]
`;

exports[`http POST http://example.com/for-good with ["<FormData> [[\\"for\\",\\"the throne\\"]]"] should work 1`] = `
Array [
  "http://example.com/for-good",
  Object {
    "body": "<FormData> [[\\"for\\",\\"the throne\\"]]",
    "headers": Headers {
      "map": Object {},
    },
    "method": "POST",
  },
]
`;

exports[`http POST http://example.com/for-good with ["for=the-throne"] should work 1`] = `
Array [
  "http://example.com/for-good",
  Object {
    "body": "for=the-throne",
    "headers": Headers {
      "map": Object {
        "content-type": "application/x-www-form-urlencoded",
      },
    },
    "method": "POST",
  },
]
`;

exports[`http POST http://example.com/for-good with [{"for":"the throne"},{"headers":{"map":{"x-requested-with":"XMLHttpRequest"}}}] should work 1`] = `
Array [
  "http://example.com/for-good",
  Object {
    "body": "{\\"for\\":\\"the throne\\"}",
    "headers": Headers {
      "map": Object {
        "content-type": "application/json",
        "x-requested-with": "XMLHttpRequest",
      },
    },
    "method": "POST",
  },
]
`;

exports[`http POST http://example.com/for-good with [{"for":"the throne"},{"headers":{"x-requested-with":"XMLHttpRequest"}}] should work 1`] = `
Array [
  "http://example.com/for-good",
  Object {
    "body": "{\\"for\\":\\"the throne\\"}",
    "headers": Headers {
      "map": Object {
        "content-type": "application/json",
        "x-requested-with": "XMLHttpRequest",
      },
    },
    "method": "POST",
  },
]
`;

exports[`http POST http://example.com/for-good with [{"for":"the throne"},{"params":"<URLSearchParams> lang=zh-CN"}] should work 1`] = `
Array [
  "http://example.com/for-good?lang=zh-CN",
  Object {
    "body": "{\\"for\\":\\"the throne\\"}",
    "headers": Headers {
      "map": Object {
        "content-type": "application/json",
      },
    },
    "method": "POST",
  },
]
`;

exports[`http POST http://example.com/for-good with [{"for":"the throne"},{"params":{"lang":"zh-CN","type":null}}] should work 1`] = `
Array [
  "http://example.com/for-good?lang=zh-CN&type=&category=",
  Object {
    "body": "{\\"for\\":\\"the throne\\"}",
    "headers": Headers {
      "map": Object {
        "content-type": "application/json",
      },
    },
    "method": "POST",
  },
]
`;

exports[`http PUT http://example.com with [] should work 1`] = `
Array [
  "http://example.com",
  Object {
    "headers": undefined,
    "method": "PUT",
  },
]
`;

exports[`http requestWithBody or simpleRequest DELETE http://example.com with [] should work 1`] = `
Array [
  "http://example.com",
  Object {
    "method": "DELETE",
  },
]
`;

exports[`http requestWithBody or simpleRequest GET http://example.com with [] should work 1`] = `
Array [
  "http://example.com",
  Object {
    "method": "GET",
  },
]
`;

exports[`http requestWithBody or simpleRequest GET http://example.com with [{"params":"<URLSearchParams> lang=zh-CN"}] should work 1`] = `
Array [
  "http://example.com/?lang=zh-CN",
  Object {
    "method": "GET",
  },
]
`;

exports[`http requestWithBody or simpleRequest GET http://example.com with [{"params":{"lang":"zh-CN","type":null}}] should work 1`] = `
Array [
  "http://example.com/?lang=zh-CN&type=&category=",
  Object {
    "method": "GET",
  },
]
`;

exports[`http requestWithBody or simpleRequest GET http://example.com with [{"params":{"lang":["zh-CN","en-US"],"type":null}}] should work 1`] = `
Array [
  "http://example.com/?lang=zh-CN&lang=en-US&type=&category=",
  Object {
    "method": "GET",
  },
]
`;

exports[`http requestWithBody or simpleRequest GET http://example.com/for-good?token=secret with [{"params":{"lang":"zh-CN","type":null}}] should work 1`] = `
Array [
  "http://example.com/for-good?token=secret&lang=zh-CN&type=&category=",
  Object {
    "method": "GET",
  },
]
`;

exports[`http requestWithBody or simpleRequest HEAD http://example.com with [] should work 1`] = `
Array [
  "http://example.com",
  Object {
    "method": "HEAD",
  },
]
`;

exports[`http requestWithBody or simpleRequest PATCH http://example.com with [] should work 1`] = `
Array [
  "http://example.com",
  Object {
    "headers": undefined,
    "method": "PATCH",
  },
]
`;

exports[`http requestWithBody or simpleRequest POST http://example.com with [] should work 1`] = `
Array [
  "http://example.com",
  Object {
    "headers": undefined,
    "method": "POST",
  },
]
`;

exports[`http requestWithBody or simpleRequest POST http://example.com with [{"for":"good"}] should work 1`] = `
Array [
  "http://example.com",
  Object {
    "body": "{\\"for\\":\\"good\\"}",
    "headers": Headers {
      "map": Object {
        "content-type": "application/json",
      },
    },
    "method": "POST",
  },
]
`;

exports[`http requestWithBody or simpleRequest POST http://example.com/for-good with ["<FormData> [[\\"for\\",\\"the throne\\"]]"] should work 1`] = `
Array [
  "http://example.com/for-good",
  Object {
    "body": "<FormData> [[\\"for\\",\\"the throne\\"]]",
    "headers": Headers {
      "map": Object {},
    },
    "method": "POST",
  },
]
`;

exports[`http requestWithBody or simpleRequest POST http://example.com/for-good with ["for=the-throne"] should work 1`] = `
Array [
  "http://example.com/for-good",
  Object {
    "body": "for=the-throne",
    "headers": Headers {
      "map": Object {
        "content-type": "application/x-www-form-urlencoded",
      },
    },
    "method": "POST",
  },
]
`;

exports[`http requestWithBody or simpleRequest POST http://example.com/for-good with [{"for":"the throne"},{"headers":{"map":{"x-requested-with":"XMLHttpRequest"}}}] should work 1`] = `
Array [
  "http://example.com/for-good",
  Object {
    "body": "{\\"for\\":\\"the throne\\"}",
    "headers": Headers {
      "map": Object {
        "content-type": "application/json",
        "x-requested-with": "XMLHttpRequest",
      },
    },
    "method": "POST",
  },
]
`;

exports[`http requestWithBody or simpleRequest POST http://example.com/for-good with [{"for":"the throne"},{"headers":{"x-requested-with":"XMLHttpRequest"}}] should work 1`] = `
Array [
  "http://example.com/for-good",
  Object {
    "body": "{\\"for\\":\\"the throne\\"}",
    "headers": Headers {
      "map": Object {
        "content-type": "application/json",
        "x-requested-with": "XMLHttpRequest",
      },
    },
    "method": "POST",
  },
]
`;

exports[`http requestWithBody or simpleRequest POST http://example.com/for-good with [{"for":"the throne"},{"params":"<URLSearchParams> lang=zh-CN"}] should work 1`] = `
Array [
  "http://example.com/for-good?lang=zh-CN",
  Object {
    "body": "{\\"for\\":\\"the throne\\"}",
    "headers": Headers {
      "map": Object {
        "content-type": "application/json",
      },
    },
    "method": "POST",
  },
]
`;

exports[`http requestWithBody or simpleRequest POST http://example.com/for-good with [{"for":"the throne"},{"params":{"lang":"zh-CN","type":null}}] should work 1`] = `
Array [
  "http://example.com/for-good?lang=zh-CN&type=&category=",
  Object {
    "body": "{\\"for\\":\\"the throne\\"}",
    "headers": Headers {
      "map": Object {
        "content-type": "application/json",
      },
    },
    "method": "POST",
  },
]
`;

exports[`http requestWithBody or simpleRequest PUT http://example.com with [] should work 1`] = `
Array [
  "http://example.com",
  Object {
    "headers": undefined,
    "method": "PUT",
  },
]
`;

exports[`http should return http response object  1`] = `
Object {
  "config": Object {
    "method": "GET",
    "options": Object {
      "observe": "response",
    },
    "url": "http://example.com",
  },
  "data": Object {
    "foo": "bar",
  },
  "headers": Headers {
    "map": Object {
      "content-type": "text/plain;charset=UTF-8",
    },
  },
  "status": 200,
  "statusText": "",
}
`;

exports[`http should work with http.request 1`] = `
Array [
  "http://example.com/for-good",
  Object {
    "method": "GET",
  },
]
`;
